dvc/scm/git.py

```diff
@@ -491,3 +497,160 @@ def resolve_commit(self, rev):
+                args = [pattern]
+            else:
+                args = []
+            for line in self.repo.git.for_each_ref(
+                *args, contains=rev, format=r"%(refname)"
+            ).splitlines():
+                line = line.strip()
+                if line:
+                    yield line
+        except GitCommandError:
+            pass
+
+    @contextmanager
+    def detach_head(self, rev: Optional[str] = None):
+        """Context manager for performing detached HEAD SCM operations.
+
+        Detaches and restores HEAD similar to interactive git rebase.
+        Restore is equivalent to 'reset --soft', meaning the caller is
+        is responsible for preserving & restoring working tree state
+        (i.e. via stash) when applicable.
+
+        Yields revision of detached head.
+        """
+        if not rev:
+            rev = "HEAD"
+        orig_head = self.get_ref("HEAD", follow=False)
+        logger.debug("Detaching HEAD at '%s'", rev)
+        self.checkout(rev, detach=True)
+        try:
+            yield self.get_ref("HEAD")
+        finally:
+            prefix = os.fsdecode(self.LOCAL_BRANCH_PREFIX)
+            if orig_head.startswith(prefix):
+                orig_head = orig_head[len(prefix) :]
+            logger.debug("Restore HEAD to '%s'", orig_head)
+            self.checkout(orig_head)
+
+    @contextmanager
+    def stash_workspace(self, **kwargs):
+        """Stash restore any workspace changes.
+
+        Yields revision of the stash commit. Yields None if there were no
+        changes to stash.
+        """
+        logger.debug("Stashing workspace")
+        rev = self.stash.push(**kwargs)
+        try:
+            yield rev
+        finally:
+            if rev:
+                logger.debug("Restoring stashed workspace")
+                self.stash.pop()
+
+
+class Stash:
+    """Wrapper for representing Git stash.
+
+    Uses dulwich.stash when possible, `git stash` will be used directly for
+    operations which are not implemented in dulwich.
+    """
+
+    DEFAULT_STASH = "refs/stash"
+
+    def __init__(self, scm, ref: Optional[str] = None):
+        self.ref = ref if ref else self.DEFAULT_STASH
+        self.scm = scm
+
+    @property
+    def git(self):
+        return self.scm.repo.git
+
+    def __iter__(self):
+        yield from self.scm._stash_iter(self.ref)
+
+    def __len__(self):
+        return len(self.list())
+
+    def __getitem__(self, index):
+        return self.list()[index]
+
+    def list(self):
+        return list(iter(self))
+
+    def push(
+        self,
+        message: Optional[str] = None,
+        include_untracked: Optional[bool] = False,
+    ) -> Optional[str]:
+        if not self.scm.is_dirty(untracked_files=include_untracked):
+            logger.debug("No changes to stash")
+            return None
+
+        logger.debug("Stashing changes in '%s'", self.ref)
+        rev, reset = self.scm._stash_push(  # pylint: disable=protected-access
```

Comment:
The backend `_stash...` functions are marked as protected to make it obvious that when using git stash you should be using this higher level `Stash` class and not the low level backend functions.